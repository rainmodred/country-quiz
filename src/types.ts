interface Currency {
  code: string | null;
  name: string | null;
  symbol: string | null;
}

interface Language {
  iso639_1: string | null; // eslint-disable-line
  iso639_2: string | null; // eslint-disable-line
  name: string;
  nativeName: string;
}

interface Translations {
  de: string | null;
  es: string | null;
  fr: string | null;
  ja: string | null;
  it: string | null;
  br: string | null;
  pt: string | null;
  nl: string | null;
  hr: string | null;
  fa: string | null;
}

interface RegionalBlocs {
  acronym: string;
  name: string;
  otherAcronyms: string[];
  otherNames: string[];
}

interface Country {
  name: string;
  topLevelDomain: string[];
  alpha2Code: string;
  alpha3Code: string;
  callingCodes: string[];
  capital: string;
  altSpellings: string[];
  region: string;
  subregion: string;
  population: number;
  latlng: number[];
  demonym: string;
  area: number | null;
  gini: number | null;
  timezones: string[];
  borders: string[];
  nativeName: string;
  numericCode: string | null;
  currencies: Currency[];
  languages: Language[];
  translations: Translations;
  flag: string;
  regionalBlocs: RegionalBlocs[];
  cioc: string | null;
}

interface Answer {
  text: string;
  correct: boolean;
}

interface Question {
  type: 'flag' | 'capital';
  name: string;
  text: string;
  asnwers: Answer[];
  flag: string;
}

type Questions = Question[];
type Answers = Answer[];
type Countries = Country[];

export type { Country, Countries, Question, Questions, Answer, Answers };
