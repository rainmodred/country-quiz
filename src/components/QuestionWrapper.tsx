/** @jsxImportSource @emotion/react */
import { css } from '@emotion/react';
import { useEffect, useState } from 'react';

import Button from './Button';
import Option from './Option';
import { Answer, Question } from '../types';
import { colors } from '../styles/colors';

interface QuestionProps {
  question: Question;
  questionIndex: number;
  totalQuestions: number;
  onAnswerClick: (answer: boolean) => void;
  onNextClick: () => void;
}

export default function QuestionWrapper({
  question,
  questionIndex,
  totalQuestions,
  onAnswerClick,
  onNextClick,
}: QuestionProps) {
  const [currentAnswer, setCurrentAnswer] = useState<Answer | null>(null);
  const [disabled, setDisabled] = useState(false);

  const { type, flag, text, asnwers } = question;

  function handleOptionClick(answer: Answer) {
    setDisabled(true);
    setCurrentAnswer(answer);
    onAnswerClick(answer.correct);
  }

  useEffect(() => {
    setDisabled(false);
  }, [question]);

  const isWrong = disabled && !currentAnswer?.correct;

  return (
    <div
      css={css`
        padding: 68px 32px 64px 32px;
        display: grid;
        row-gap: 25px;
        width: 100%;
        @media (max-width: 768px) {
          row-gap: 10px;
          padding: 65px 25px 25px 25px;
          padding-top: ${type === 'flag' ? '20px' : '65px'};
        }
      `}
    >
      {type === 'flag' && (
        <img
          src={flag}
          alt="flag"
          css={css`
            width: 84px;
            filter: drop-shadow(0px 4px 24px rgba(0, 0, 0, 0.1));
            @media (max-width: 768px) {
              padding-top: 20px;
            }
          `}
        />
      )}
      <h3
        css={css`
          color: ${colors.colorQuestionText};
          font-size: 24px;
          line-height: 36px;
          font-weight: 700;
          @media (max-width: 768px) {
            font-size: 24px;
            line-height: 34px;
          }
        `}
      >
        {text}
      </h3>
      {asnwers.map(answer => (
        <Option
          answer={answer}
          wrong={isWrong && answer.text === currentAnswer?.text}
          disabled={disabled}
          onClick={handleOptionClick}
          key={answer.text}
        />
      ))}

      <div
        css={css`
          display: flex;
          justify-content: space-between;
          align-items: center;
          height: 56px;
        `}
      >
        <span
          css={css`
            font-size: 14px;
            margin-left: 5px;
          `}
        >
          {questionIndex + 1} / {totalQuestions}
        </span>
        {disabled && (
          <Button primary onClick={onNextClick}>
            Next
          </Button>
        )}
      </div>
    </div>
  );
}
