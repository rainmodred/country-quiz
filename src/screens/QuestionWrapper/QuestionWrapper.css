.question {
  padding: 68px 32px 64px 32px;
  display: grid;
  row-gap: 25px;
  width: 100%;
}

.question__flag {
  width: 84px;
  filter: drop-shadow(0px 4px 24px rgba(0, 0, 0, 0.1));
}

.question__header {
  color: var(--color-question-text);
  font-size: 24px;
  line-height: 36px;
  font-weight: 700;
}

.question__footer {
  display: flex;
  justify-content: space-between;
  align-items: center;
  height: 56px;
}

.question__count {
  font-size: 14px;
  margin-left: 5px;
}

@media (max-width: 768px) {
  .question {
    row-gap: 10px;
    padding: 65px 25px 25px 25px;
  }

  .question--flag {
    padding-top: 20px;
  }

  .question__header {
    font-size: 24px;
    line-height: 34px;
  }
}
