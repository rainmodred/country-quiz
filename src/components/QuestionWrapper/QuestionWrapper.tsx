import { useEffect, useState } from 'react';
import Button from '../Button/Button';
import Option from '../Option/Option';
import { Answer, Question } from '../../types';
import './QuestionWrapper.css';

interface QuestionProps {
  question: Question;
  questionIndex: number;
  totalQuestions: number;
  onAnswerClick: (answer: boolean) => void;
  onNextClick: () => void;
}

export default function QuestionWrapper({
  question,
  questionIndex,
  totalQuestions,
  onAnswerClick,
  onNextClick,
}: QuestionProps) {
  const [currentAnswer, setCurrentAnswer] = useState<Answer | null>(null);
  const [disabled, setDisabled] = useState(false);

  function handleOptionClick(answer: Answer) {
    setDisabled(true);
    setCurrentAnswer(answer);
    onAnswerClick(answer.correct);
  }

  useEffect(() => {
    setDisabled(false);
  }, [question]);

  const isWrong = disabled && !currentAnswer?.correct;

  const classes = `${
    question.type === 'flag' ? 'question question--flag' : 'question'
  }`;
  return (
    <div className={classes}>
      {question.type === 'flag' && (
        <img src={question.flag} alt="flag" className="question__flag" />
      )}
      <h3 className="question__header">{question.text}</h3>
      {question.asnwers.map(answer => (
        <Option
          answer={answer}
          wrong={isWrong && answer.text === currentAnswer?.text}
          disabled={disabled}
          onClick={handleOptionClick}
          key={answer.text}
        />
      ))}

      <div className="question__footer">
        <span className="question__count">
          {questionIndex + 1} / {totalQuestions}
        </span>
        {disabled && (
          <Button primary onClick={onNextClick}>
            Next
          </Button>
        )}
      </div>
    </div>
  );
}
