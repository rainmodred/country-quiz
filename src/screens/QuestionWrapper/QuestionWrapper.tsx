import { useState } from 'react';
import Button from '../../components/Button/Button';
import Option from '../../components/Option/Option';
import { Question } from '../../types';

import './QuestionWrapper.css';

interface QuestionProps {
  question: Question;
  disabled: boolean;
  onAnswerClick: (answer: boolean) => void;
  onNextClick: () => void;
}

export default function QuestionWrapper({
  question,
  disabled,
  onAnswerClick,
  onNextClick,
}: QuestionProps) {
  const [currentAnswer, setCurrentAnswer] = useState('');
  const correctAnswer = question?.asnwers.filter(ans => ans.correct)[0].text;

  function handleOptionClick(text: string) {
    setCurrentAnswer(text);
    onAnswerClick(text === correctAnswer);
  }

  // WTF IS THIS
  const isWrong =
    disabled && currentAnswer !== '' && currentAnswer !== correctAnswer;

  if (!question) {
    return <p>Loading...</p>;
  }

  return (
    <div className="question">
      {question.type === 'flag' && (
        <img src={question.flag} alt="flag" className="question__flag" />
      )}
      <h3 className="question__header">{question.text}</h3>
      {question.asnwers.map(answer => (
        <Option
          text={answer.text}
          correct={answer.correct}
          error={isWrong && answer.text === currentAnswer}
          disabled={disabled}
          onClick={handleOptionClick}
          key={answer.text}
        />
      ))}

      {disabled && (
        <Button primary onClick={onNextClick}>
          Next
        </Button>
      )}
    </div>
  );
}
