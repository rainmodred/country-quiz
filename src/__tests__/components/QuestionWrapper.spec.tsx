import { render } from '@testing-library/react';
import user from '@testing-library/user-event';
import { Question } from 'types';

import QuestionWrapper from '../../components/QuestionWrapper/QuestionWrapper';

const flagQuestion: Question = {
  type: 'flag',
  name: 'Germany',
  flag: 'https://restcountries.eu/data/deu.svg',
  text: 'Which country does this flag belong to?',
  asnwers: [
    {
      variant: 'A',
      text: 'Basseterre',
      correct: false,
    },
    {
      variant: 'B',
      text: 'Maputo',
      correct: false,
    },
    {
      variant: 'C',
      text: 'Germany',
      correct: true,
    },
    {
      variant: 'D',
      text: 'Dhaka',
      correct: false,
    },
  ],
};

const capitalQuestion: Question = {
  type: 'capital',
  name: 'Switzerland',
  flag: 'https://restcountries.eu/data/che.svg',
  text: 'Bern is the capital of',
  asnwers: [
    {
      variant: 'A',
      text: 'Switzerland',
      correct: true,
    },
    {
      variant: 'B',
      text: 'Lisbon',
      correct: false,
    },
    {
      variant: 'C',
      text: 'Brasília',
      correct: false,
    },
    {
      variant: 'D',
      text: 'Quito',
      correct: false,
    },
  ],
};

describe('QuestionWrapper', () => {
  const handleAnswerClick = jest.fn();
  const handleNextClick = jest.fn();
  beforeEach(() => {
    handleAnswerClick.mockClear();
    handleNextClick.mockClear();
  });

  it('renders with flagQuestion', () => {
    const { container, getByAltText, getByText, queryByText } = render(
      <QuestionWrapper
        question={flagQuestion}
        questionIndex={0}
        totalQuestions={10}
        onAnswerClick={handleAnswerClick}
        onNextClick={handleNextClick}
      />,
    );

    expect(getByText(/Which country does this flag belong to?/i)).toBeTruthy();
    expect(getByAltText(/flag/i)).toBeTruthy();
    expect(getByText(/Basseterre/i)).toBeTruthy();
    expect(getByText(/Maputo/i)).toBeTruthy();
    expect(getByText(/Germany/i)).toBeTruthy();
    expect(getByText(/Dhaka/i)).toBeTruthy();
    expect(queryByText(/next/i)).toBeNull();
    expect(container).toMatchSnapshot();
  });

  it('renders with capitalQuestion', () => {
    const { container, getByText, queryByText, queryByAltText } = render(
      <QuestionWrapper
        question={capitalQuestion}
        questionIndex={0}
        totalQuestions={10}
        onAnswerClick={handleAnswerClick}
        onNextClick={handleNextClick}
      />,
    );

    expect(getByText(/Bern is the capital of/i)).toBeTruthy();
    expect(queryByAltText(/flag/i)).toBeNull();
    expect(getByText(/Switzerland/i)).toBeTruthy();
    expect(getByText(/Lisbon/i)).toBeTruthy();
    expect(getByText(/Brasília/i)).toBeTruthy();
    expect(getByText(/Quito/i)).toBeTruthy();
    expect(queryByText(/next/i)).toBeNull();
    expect(container).toMatchSnapshot();
  });

  it('renders correct answer', () => {
    const {
      container,
      getByText,
      queryByText,
      queryByTitle,
      queryByAltText,
    } = render(
      <QuestionWrapper
        question={capitalQuestion}
        questionIndex={0}
        totalQuestions={10}
        onAnswerClick={handleAnswerClick}
        onNextClick={handleNextClick}
      />,
    );
    const correctAnswer = getByText(/Switzerland/i);

    expect(getByText(/Bern is the capital of/i)).toBeTruthy();
    expect(queryByAltText(/flag/i)).toBeNull();
    expect(correctAnswer).toBeTruthy();
    expect(getByText(/Lisbon/i)).toBeTruthy();
    expect(getByText(/Brasília/i)).toBeTruthy();
    expect(getByText(/Quito/i)).toBeTruthy();
    expect(queryByText(/next/i)).toBeNull();
    user.click(correctAnswer);
    expect(queryByTitle(/checkIcon/i)).toBeTruthy();
    expect(queryByText(/next/i)).toBeTruthy();
    expect(container).toMatchSnapshot();
  });

  it('renders wrong answer', () => {
    const {
      container,
      getByText,
      queryByText,
      queryByTitle,
      queryByAltText,
    } = render(
      <QuestionWrapper
        question={capitalQuestion}
        questionIndex={0}
        totalQuestions={10}
        onAnswerClick={handleAnswerClick}
        onNextClick={handleNextClick}
      />,
    );
    const correctAnswer = getByText(/Switzerland/i);
    const wrongAnswer = getByText(/Lisbon/i);

    expect(getByText(/Bern is the capital of/i)).toBeTruthy();
    expect(queryByAltText(/flag/i)).toBeNull();
    expect(correctAnswer).toBeTruthy();
    expect(wrongAnswer).toBeTruthy();
    expect(getByText(/Brasília/i)).toBeTruthy();
    expect(getByText(/Quito/i)).toBeTruthy();
    expect(queryByText(/next/i)).toBeNull();
    user.click(wrongAnswer);
    expect(queryByTitle(/checkIcon/i)).toBeTruthy();
    expect(queryByTitle(/closeIcon/i)).toBeTruthy();
    expect(queryByText(/next/i)).toBeTruthy();
    expect(container).toMatchSnapshot();
  });

  it('renders question count', () => {
    const { getByText } = render(
      <QuestionWrapper
        question={capitalQuestion}
        questionIndex={5}
        totalQuestions={10}
        onAnswerClick={handleAnswerClick}
        onNextClick={handleNextClick}
      />,
    );

    expect(getByText(/6 \/ 10/i)).toBeTruthy();
  });
});
