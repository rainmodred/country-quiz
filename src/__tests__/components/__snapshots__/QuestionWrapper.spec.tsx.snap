// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionWrapper renders correct answer 1`] = `
<div>
  <div
    class="css-11ud1ff-QuestionWrapper"
  >
    <h3
      class="css-1cjrmqs-QuestionWrapper"
    >
      Bern is the capital of
    </h3>
    <button
      class="css-4ptem4-Option"
      type="button"
    >
      <span
        class="css-13f8g8s-Option"
      >
        A
      </span>
      <span
        class="css-oh1baf-Option"
      >
        Switzerland
      </span>
      <svg
        fill="currentColor"
        height="20"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 1024 1024"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <title>
          checkIcon
        </title>
        <path
          d="M699 353h-46.9c-10.2 0-19.9 4.9-25.9 13.3L469 584.3l-71.2-98.8c-6-8.3-15.6-13.3-25.9-13.3H325c-6.5 0-10.3 7.4-6.5 12.7l124.6 172.8a31.8 31.8 0 0 0 51.7 0l210.6-292c3.9-5.3.1-12.7-6.4-12.7z"
        />
        <path
          d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
        />
      </svg>
    </button>
    <button
      class="css-5v9hz0-Option"
      type="button"
    >
      <span
        class="css-13f8g8s-Option"
      >
        B
      </span>
      <span
        class="css-oh1baf-Option"
      >
        Lisbon
      </span>
    </button>
    <button
      class="css-5v9hz0-Option"
      type="button"
    >
      <span
        class="css-13f8g8s-Option"
      >
        C
      </span>
      <span
        class="css-oh1baf-Option"
      >
        Brasília
      </span>
    </button>
    <button
      class="css-5v9hz0-Option"
      type="button"
    >
      <span
        class="css-13f8g8s-Option"
      >
        D
      </span>
      <span
        class="css-oh1baf-Option"
      >
        Quito
      </span>
    </button>
    <div
      class="css-ipgmsr-QuestionWrapper"
    >
      <span
        class="css-17tonp7-QuestionWrapper"
      >
        1
         / 
        10
      </span>
      <button
        class="css-inaz0x-Button"
        type="button"
      >
        Next
      </button>
    </div>
  </div>
</div>
`;

exports[`QuestionWrapper renders with capitalQuestion 1`] = `
<div>
  <div
    class="css-11ud1ff-QuestionWrapper"
  >
    <h3
      class="css-1cjrmqs-QuestionWrapper"
    >
      Bern is the capital of
    </h3>
    <button
      class="css-1yoohz3-Option"
      type="button"
    >
      <span
        class="css-13f8g8s-Option"
      >
        A
      </span>
      <span
        class="css-oh1baf-Option"
      >
        Switzerland
      </span>
    </button>
    <button
      class="css-1yoohz3-Option"
      type="button"
    >
      <span
        class="css-13f8g8s-Option"
      >
        B
      </span>
      <span
        class="css-oh1baf-Option"
      >
        Lisbon
      </span>
    </button>
    <button
      class="css-1yoohz3-Option"
      type="button"
    >
      <span
        class="css-13f8g8s-Option"
      >
        C
      </span>
      <span
        class="css-oh1baf-Option"
      >
        Brasília
      </span>
    </button>
    <button
      class="css-1yoohz3-Option"
      type="button"
    >
      <span
        class="css-13f8g8s-Option"
      >
        D
      </span>
      <span
        class="css-oh1baf-Option"
      >
        Quito
      </span>
    </button>
    <div
      class="css-ipgmsr-QuestionWrapper"
    >
      <span
        class="css-17tonp7-QuestionWrapper"
      >
        1
         / 
        10
      </span>
    </div>
  </div>
</div>
`;

exports[`QuestionWrapper renders with flagQuestion 1`] = `
<div>
  <div
    class="css-1nzzrsz-QuestionWrapper"
  >
    <img
      alt="flag"
      class="css-8dd1yu-QuestionWrapper"
      src="https://restcountries.eu/data/deu.svg"
    />
    <h3
      class="css-1cjrmqs-QuestionWrapper"
    >
      Which country does this flag belong to?
    </h3>
    <button
      class="css-1yoohz3-Option"
      type="button"
    >
      <span
        class="css-13f8g8s-Option"
      >
        A
      </span>
      <span
        class="css-oh1baf-Option"
      >
        Basseterre
      </span>
    </button>
    <button
      class="css-1yoohz3-Option"
      type="button"
    >
      <span
        class="css-13f8g8s-Option"
      >
        B
      </span>
      <span
        class="css-oh1baf-Option"
      >
        Maputo
      </span>
    </button>
    <button
      class="css-1yoohz3-Option"
      type="button"
    >
      <span
        class="css-13f8g8s-Option"
      >
        C
      </span>
      <span
        class="css-oh1baf-Option"
      >
        Germany
      </span>
    </button>
    <button
      class="css-1yoohz3-Option"
      type="button"
    >
      <span
        class="css-13f8g8s-Option"
      >
        D
      </span>
      <span
        class="css-oh1baf-Option"
      >
        Dhaka
      </span>
    </button>
    <div
      class="css-ipgmsr-QuestionWrapper"
    >
      <span
        class="css-17tonp7-QuestionWrapper"
      >
        1
         / 
        10
      </span>
    </div>
  </div>
</div>
`;

exports[`QuestionWrapper renders wrong answer 1`] = `
<div>
  <div
    class="css-11ud1ff-QuestionWrapper"
  >
    <h3
      class="css-1cjrmqs-QuestionWrapper"
    >
      Bern is the capital of
    </h3>
    <button
      class="css-4ptem4-Option"
      type="button"
    >
      <span
        class="css-13f8g8s-Option"
      >
        A
      </span>
      <span
        class="css-oh1baf-Option"
      >
        Switzerland
      </span>
      <svg
        fill="currentColor"
        height="20"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 1024 1024"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <title>
          checkIcon
        </title>
        <path
          d="M699 353h-46.9c-10.2 0-19.9 4.9-25.9 13.3L469 584.3l-71.2-98.8c-6-8.3-15.6-13.3-25.9-13.3H325c-6.5 0-10.3 7.4-6.5 12.7l124.6 172.8a31.8 31.8 0 0 0 51.7 0l210.6-292c3.9-5.3.1-12.7-6.4-12.7z"
        />
        <path
          d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
        />
      </svg>
    </button>
    <button
      class="css-1wf7kng-Option"
      type="button"
    >
      <span
        class="css-13f8g8s-Option"
      >
        B
      </span>
      <span
        class="css-oh1baf-Option"
      >
        Lisbon
      </span>
      <svg
        fill="currentColor"
        height="20"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 1024 1024"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <title>
          closeIcon
        </title>
        <path
          d="M685.4 354.8c0-4.4-3.6-8-8-8l-66 .3L512 465.6l-99.3-118.4-66.1-.3c-4.4 0-8 3.5-8 8 0 1.9.7 3.7 1.9 5.2l130.1 155L340.5 670a8.32 8.32 0 0 0-1.9 5.2c0 4.4 3.6 8 8 8l66.1-.3L512 564.4l99.3 118.4 66 .3c4.4 0 8-3.5 8-8 0-1.9-.7-3.7-1.9-5.2L553.5 515l130.1-155c1.2-1.4 1.8-3.3 1.8-5.2z"
        />
        <path
          d="M512 65C264.6 65 64 265.6 64 513s200.6 448 448 448 448-200.6 448-448S759.4 65 512 65zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
        />
      </svg>
    </button>
    <button
      class="css-5v9hz0-Option"
      type="button"
    >
      <span
        class="css-13f8g8s-Option"
      >
        C
      </span>
      <span
        class="css-oh1baf-Option"
      >
        Brasília
      </span>
    </button>
    <button
      class="css-5v9hz0-Option"
      type="button"
    >
      <span
        class="css-13f8g8s-Option"
      >
        D
      </span>
      <span
        class="css-oh1baf-Option"
      >
        Quito
      </span>
    </button>
    <div
      class="css-ipgmsr-QuestionWrapper"
    >
      <span
        class="css-17tonp7-QuestionWrapper"
      >
        1
         / 
        10
      </span>
      <button
        class="css-inaz0x-Button"
        type="button"
      >
        Next
      </button>
    </div>
  </div>
</div>
`;
